src/search/params.cc

```diff
@@ -30,16 +30,67 @@ SearchQuery::clear()
   void
 from_json( const nlohmann::json & jfrom, SearchQuery & qry )
 {
-  pkgdb::from_json( jfrom, dynamic_cast<pkgdb::PkgDescriptorBase &>( qry ) );
-  try { jfrom.at( "partialMatch" ).get_to( qry.partialMatch ); }
-  catch( const nlohmann::json::out_of_range & ) {}
+  auto getOrFail = [&]( const std::string    & key
+                      , const nlohmann::json & from
+                      , auto                 & sink
+                      )
+    {
+      try
+        {
+          from.get_to( sink );
+        }
+      catch( const nlohmann::json::exception & err )
+        {
+          throw FloxException(
+            "Failed to parse search query field: 'query." + key + "':\n  " +
+            err.what()
+          );
+        }
+      catch( ... )
+        {
+          throw FloxException(
+            "Failed to parse search query field: 'query." + key + "'."
+          );
+        }
+    };
+
+  for ( const auto & [key, value] : jfrom.items() )
+    {
+      if ( key == "name" )
+        {
+          getOrFail( key, value, qry.name );
+        }
+      else if ( key == "pname" )
+        {
+          getOrFail( key, value, qry.pname );
+        }
+      else if ( key == "version" )
+        {
+          getOrFail( key, value, qry.version );
+        }
+      else if ( key == "semver" )
+        {
+          getOrFail( key, value, qry.semver );
+        }
+      else if ( key == "match" )
+        {
+          getOrFail( key, value, qry.partialMatch );
+        }
+      else
+        {
+          throw FloxException(
+            "Unrecognized search query key: 'query." + key + "'."
```

Comment:
Is there no easy way to do this with `nlohmann`? Just looked at the macros briefly, and it looks like they can do the equivalent of the `getOrFail` stuff, but maybe not throwing for an unrecognized field?

Reply:
Yeah they can if you use the `NLOHMANN_DEFINE_TYPE_*` macros, but those unfortunately will not get the "Unrecognized field" error.

With that in mind I needed to do a `from_json` _by hand_.

The `getOrFail` might not be necessary, but I wanted the messages to read _sort of like an attribute path_ so "failed to parse `query.name`: <nlohmann error messag>" instead of just the plain nlohmann error since it lacks context.